java/src/main/java/com/logicalclocks/hsfs/engine/hudi/HudiEngine.java

```diff
@@ -233,9 +247,30 @@ private FeatureGroupCommit getLastCommitMetadata(SparkSession sparkSession, Stri
     return hudiArgs;
   }
 
-  @SneakyThrows
-  public String timeStampToHudiFormat(Long commitedOnTimeStamp) {
-    Date commitedOnDate = new Timestamp(commitedOnTimeStamp);
-    return dateFormat.format(commitedOnDate);
+  public void streamToHoodieTable(SparkSession sparkSession, StreamFeatureGroup streamFeatureGroup,
+                                  Map<String, String> writeOptions) throws Exception {
+
+    Map<String, String> hudiWriteOpts = setupHudiWriteOpts(streamFeatureGroup, HudiOperationType.UPSERT, writeOptions);
+    hudiWriteOpts.put("projectId", String.valueOf(streamFeatureGroup.getFeatureStore().getProjectId()));
+    hudiWriteOpts.put("featureStoreName", streamFeatureGroup.getFeatureStore().getName());
+    hudiWriteOpts.put("featureGroupName", streamFeatureGroup.getName());
+    hudiWriteOpts.put("featureGroupVersion", String.valueOf(streamFeatureGroup.getVersion()));
+    hudiWriteOpts.put(HUDI_TABLE_NAME, utils.getFgName(streamFeatureGroup));
+    hudiWriteOpts.put(HUDI_BASE_PATH, streamFeatureGroup.getLocation());
+    hudiWriteOpts.put(CHECKPOINT_PROVIDER_PATH_PROP, streamFeatureGroup.getLocation());
+    hudiWriteOpts.put(HUDI_KAFKA_TOPIC, streamFeatureGroup.getOnlineTopicName());
+    hudiWriteOpts.put(FEATURE_GROUP_SCHEMA, streamFeatureGroup.getAvroSchema());
```

Comment:
This will give you the full avro schema, e.g. with the complex types. I am unsure if it will work to deserialize it with that directly. or if you need to deserialize twice.